Nginx 1.23.0 header changes

Looks like nginx 1.23.0 header changes https://nginx.org/en/CHANGES makes this srcache-nginx-module module incompatible now

```
        ../srcache-nginx-module-0.32/src/ngx_http_srcache_util.c
ccache gcc -c -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g -I/usr/local/zlib-cf/include -I/usr/local/nginx-dep/include -m64 -march=x86-64 -mavx -mavx2 -mpclmul -msse4 -msse4.1 -msse4.2 -DTCP_FASTOPEN=23 -falign-functions=32 -g -O3 -Wno-strict-aliasing -fstack-protector-strong -fuse-ld=gold --param=ssp-buffer-size=4 -Wformat -Werror=format-security -Wno-pointer-sign  -Wimplicit-fallthrough=0 -Wno-missing-profile -Wno-implicit-function-declaration -Wno-int-conversion -Wno-unused-result -Wno-unused-result -Wno-vla-parameter -fcode-hoisting -Wp,-D_FORTIFY_SOURCE=2 -Wno-deprecated-declarations -Wno-deprecated-declarations -DNDK_SET_VAR -DNDK_UPSTREAM_LIST  -I src/core -I src/event -I src/event/modules -I src/os/unix -I ../nginx-rtmp-module -I ../ngx_brotli/deps/brotli/c/include -I ../njs/nginx/../src -I ../njs/nginx/../build -I ../ngx_devel_kit-0.3.0/objs -I objs/addon/ndk -I ../ngx_devel_kit-0.3.0/src -I ../ngx_devel_kit-0.3.0/objs -I objs/addon/ndk -I ../openssl-1.1.1p/.openssl/include -I ../zlib-cloudflare-1.3.0 -I objs -I src/http -I src/http/modules -I src/http/v2 -I ../ngx_devel_kit-0.3.0/src \
        -o objs/addon/src/ngx_http_srcache_var.o \
        ../srcache-nginx-module-0.32/src/ngx_http_srcache_var.c
../ngx_http_redis-0.3.7/ngx_http_redis_module.c: In function ‘ngx_http_redis_process_header’:
../ngx_http_redis-0.3.7/ngx_http_redis_module.c:530:26: error: ‘ngx_http_upstream_headers_in_t’ has no member named ‘content_encoding’
  530 |             u->headers_in.content_encoding = h;
      |                          ^
make[1]: *** [objs/Makefile:1844: objs/addon/ngx_http_redis-0.3.7/ngx_http_redis_module.o] Error 1
make[1]: *** Waiting for unfinished jobs....
../srcache-nginx-module-0.32/src/ngx_http_srcache_util.c: In function ‘ngx_http_srcache_response_no_cache’:
../srcache-nginx-module-0.32/src/ngx_http_srcache_util.c:558:39: error: ‘r->headers_out.cache_control’ is a pointer; did you mean to use ‘->’?
  558 |     ccp = r->headers_out.cache_control.elts;
      |                                       ^
      |                                       ->
../srcache-nginx-module-0.32/src/ngx_http_srcache_util.c:564:49: error: ‘r->headers_out.cache_control’ is a pointer; did you mean to use ‘->’?
  564 |     for (i = 0; i < r->headers_out.cache_control.nelts; i++) {
      |                                                 ^
      |                                                 ->
make[1]: *** [objs/Makefile:1893: objs/addon/src/ngx_http_srcache_util.o] Error 1
make[1]: Leaving directory '/svr-setup/nginx-1.23.0'
make: *** [Makefile:10: build] Error 2
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -25,6 +25,7 @@ env:
   matrix:
     - NGINX_VERSION=1.17.8
     - NGINX_VERSION=1.19.9
+    - NGINX_VERSION=1.23.0
 
 services:
  - memcache

--- a/src/ngx_http_srcache_headers.c
+++ b/src/ngx_http_srcache_headers.c
@@ -256,9 +256,17 @@ static ngx_int_t
 ngx_http_srcache_process_multi_header_lines(ngx_http_request_t *r,
     ngx_table_elt_t *h, ngx_uint_t offset)
 {
+#if defined(nginx_version) && nginx_version < 1023000
     ngx_array_t      *pa;
+#endif
     ngx_table_elt_t  *ho, **ph;
 
+#if defined(nginx_version) && nginx_version >= 1023000
+    ph = (ngx_table_elt_t **) ((char *) &r->headers_out + offset);
+    while (*ph) {
+      ph = &(*ph)->next;
+    }
+#else
     pa = (ngx_array_t *) ((char *) &r->headers_out + offset);
 
     if (pa->elts == NULL) {
@@ -272,6 +280,7 @@ ngx_http_srcache_process_multi_header_lines(ngx_http_request_t *r,
     if (ph == NULL) {
         return NGX_ERROR;
     }
+#endif
 
     ho = ngx_list_push(&r->headers_out.headers);
     if (ho == NULL) {
@@ -280,6 +289,9 @@ ngx_http_srcache_process_multi_header_lines(ngx_http_request_t *r,
 
     *ho = *h;
     *ph = ho;
+#if defined(nginx_version) && nginx_version >= 1023000
+    ho->next = NULL;
+#endif
 
     return NGX_OK;
 }

--- a/src/ngx_http_srcache_util.c
+++ b/src/ngx_http_srcache_util.c
@@ -546,15 +546,34 @@ ngx_int_t
 ngx_http_srcache_response_no_cache(ngx_http_request_t *r,
     ngx_http_srcache_loc_conf_t *conf, ngx_http_srcache_ctx_t *ctx)
 {
-    ngx_table_elt_t   **ccp;
     ngx_table_elt_t    *h;
+#if defined(nginx_version) && nginx_version >= 1023000
+    ngx_table_elt_t    *cc;
+#else
+    ngx_table_elt_t   **ccp;
     ngx_uint_t          i;
+#endif
     u_char             *p, *last;
     ngx_int_t           n;
     time_t              expires;
 
     dd("checking response cache control settings");
 
+#if defined(nginx_version) && nginx_version >= 1023000
+    cc = r->headers_out.cache_control;
+
+    if (cc == NULL) {
+        goto check_expires;
+    }
+
+    for (cc = cc->next; cc; cc = cc->next) {
+        if (!cc->hash) {
+            continue;
+        }
+
+        p = cc->value.data;
+        last = p + cc->value.len;
+#else
     ccp = r->headers_out.cache_control.elts;
 
     if (ccp == NULL) {
@@ -568,6 +587,7 @@ ngx_http_srcache_response_no_cache(ngx_http_request_t *r,
 
         p = ccp[i]->value.data;
         last = p + ccp[i]->value.len;
+#endif
 
         if (!conf->store_private
             && ngx_strlcasestrn(p, last, (u_char *) "private", 7 - 1) != NULL)

--- a/src/ngx_http_srcache_util.c
+++ b/src/ngx_http_srcache_util.c
@@ -566,7 +566,7 @@ ngx_http_srcache_response_no_cache(ngx_http_request_t *r,
         goto check_expires;
     }
 
-    for (cc = cc->next; cc; cc = cc->next) {
+    for (; cc; cc = cc->next) {
         if (!cc->hash) {
             continue;
         }

